ml_in_a_box.sh

```diff
@@ -1,283 +1,104 @@
-# Install software for Ubuntu 20.04 ML-in-a-Box template VM on Paperspace Core
-#
-# Uses VM supplied by engineering
-# After running, VM is templatized by them
-# Script requires sudo access
-# Installs go to standard locations, not ~/src as in previous 18.04 ML-in-a-Box template from 2018
-# (except for Miniconda, required for NVidia RAPIDS)
-# Script assumes a fresh copy of the VM, with the software to be installed not already present
-# Uses a mix of install methods: Debian repo, apt-get, pip, Miniconda, depending on software
-# Some pip installs have alternative apt-get installs, but pip versions for Python/ML tend to be newer
-# See under the individual software installs below for other detail comments
-#
-# See also
-#
-# Paperspace-internal GitHub repo for this ML-in-a-Box setup: https://github.com/Paperspace/ml-in-a-box
-#
-# Last updated: Oct 04th 2021
+#!/usr/bin/env bash
 
-### TODO ###
-
-# These should be done before the VM is released
-
-# Engineering QA the GPU/CUDA/NVidia setup, especially w.r.t.
-#  - A100 GPU & CUDA so "it just works"
-#  - Versioning (most things here are fixed version, some are not; ones not have no obvious way to specify a version)
-#  - Security (we're using installed repos with arbitrary code)
-#  - Run script in one go as the superuser? (commands require sudo; VM doesn't let me be SU)
-#  - Licensing OK (looks good to me but I am not a lawyer)
-#  - Test out final VM when built
-
-# Improvements
-
-# These improve the script but are not vital
-
-# Add PyTorch 1.9/1.10 when CUDA 11.4 is supported
-# CUDA toolkit install is not verified. Later steps use it but there is a verify install using manual steps.
-# Put NumPy back to 1.21.2 after TensorFlow 2.5.0 downgrades it: pip install numpy 1.21.2 again, and then check TensorFlow still works
-# Unclear what the lib errors are when TF is asked to show the GPU devices after install
-# NVidia RAPIDS requires own virtualenv (conda, package installs, etc.), which is not ideal when nothing else installed needs it
-
-# Working directory for script
-
-mkdir ~/src
-cd ~/src
-
-
-# GPU
-# ---
-
-# - NVidia Driver 470.63.01 -
-
-# VM is set up by engineering with latest driver as of Sep 15th 2021: 470.63.01
-
-# - CUDA 11.4 -
-
-# CUDA 11.4 is present on the VM as supplied by engineering
-
-# - cuDNN 8.2.4.15-1+cuda11.4 -
-
-# Following https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html section 2.3.4.1. Ubuntu Network Installation
+sudo apt-get update
+sudo apt-get -y upgrade
+sudo apt-get -y install \
+  gcc \
+  make \
+  pkg-config \
+  apt-transport-https \
+  ca-certificates
+
+# Blacklist nouveau and rebuild kernel initramfs
```

Comment:
My initial research and understanding of Nouveau is that it's an open source NVIDIA driver. Is that correct?
If so, is there more commentary you can provide regarding how it might conflict with the NVIDIA driver?
I am wondering if it is enabled by default on our XenServer (RHEL-based) metal hosts and if it might be related to the PCI passthrough issues we've been having with the Ampere cards.

Reply:
Sure! So Nouveau, you're totally right - it's an OSS driver for much functionality for many NVIDIA GPUs. It comes installed by default on fresh Ubuntu images and is mutually exclusive to the proprietary NVIDIA driver. Notably, being an unofficial driver, it tends to lag behind in support for newer features and cards versus the proprietary NVIDIA drivers. (Ref. the Nouveau feature matrix for more info: https://nouveau.freedesktop.org/FeatureMatrix.html) My general experience has been that the proprietary drivers are preferred for HPC environments where you want some better guarantees of everything playing nicely among the cards, the ML frameworks, and the intermediary CUDA / cuDNN / etc. drivers